### ISSUE TITLE
CentOS 7 efforts tracking

### ISSUE THREAD

**Issue by tatarsky** [frazer-lab/cluster#199]
This Git will track the plans to move to CentOS 7 in a phased way on the cluster.

Currently it will consist mostly of me checking on Lustre update options.

**Comment by tatarsky** [comment]
AHPC ticket opened to get Intel patches for either latest IEEL 2.3 or consider IEEL 3.X.
Investigating the impact of moving to 3.X so its one less thing to update later.

Believe we could deploy a CentOS 7.0 client anytime with current distribution. (Note carefully 7.0 is the first release of CentOS)

**Comment by tatarsky** [comment]
Still working on details. Possibly requires filesystem stop to update but reviewing release notes.

**Comment by tatarsky** [comment]
Moving to IEEL 3.X is a large project and probably more than we need at first. It involves a complete burn of the Lustre servers to CentOS 7 and as such there is also risk. So I would view this step as "further" than we seek at this phase and certainly not likely before we better implement some form of backup. (See #200 and perhaps other discussions).

I have asked Intel whats the most recent CentOS 7 client I can run using our existing IEEL 2.3 branch of the code if fully updated. As I feel that makes more sense at the moment in terms of risk/reward.

**Comment by hirokomatsui** [comment]
Thanks. We'll go for CentOS 7 with IEEL 2.3.

**Comment by tatarsky** [comment]
Yep. 100% on the same page there. I'm just waiting for a response as to what "revision" of CentOS 7 we can do with patches. I believe we could deploy a CentOS 7.0 (initial release) now but want to see if I can get 7.2 as its more refined than 7.0 was. 

Note clearly the Lustre servers remain 6.X with IEEL 2.3. Which I think is fine. No user access to those systems. 

I'm likely going to propose as a test first one of the old nodes gets kickstarted. But lets wait to see what Intel says....

**Comment by tatarsky** [comment]
Answering part of my own question by digging around in the CentOS archives. 

The IEEL version we have now has client kernels that match CentOS 7.1.1503. Which isn't too bad. So I will likely grab that ISO and prep a Kickstart area while I wait for Intel response.

**Comment by tatarsky** [comment]
We are discussing the details of IEEL release 2.4.X.X which we believe we can hop to safety (more soon after I digest the release notes) and provides CentOS 7.2 client side support. Which is the latest CentOS IEEL supports in all versions. More details as I read and potential impact statement. 

Still working in parallel on 7.1 test system deployment. Probably not the fl-hn2 headnode yet though.

**Comment by tatarsky** [comment]
I am prototyping the CenOS 7 kickstart on old node `cn12`. I have removed it from the queues. Once I confirm a few more update path items there I will start talking schedules.

**Comment by tatarsky** [comment]
Actually quick check with @hirokomatsui. I am always of the belief that nobody is leaving anything important on nodes /scratch areas. I show that to be the case on cn12 at the moment but if anyone feels otherwise please inform me. cn12 will be completely erased after I eat lunch.

**Comment by hirokomatsui** [comment]
Yes, all the drives on the compute nodes can be cleared up.

**Comment by tatarsky** [comment]
OK. This is just to start some tests of "applications" as the old nodes don't really mount Lustre directly but could serve as places to see if there will be considerable impact to needed software. 

I am doing a few items in parallel.

**Comment by tatarsky** [comment]
Status only: working on some puppet basics to configure the system.

**Comment by tatarsky** [comment]
Forking software validation conversation into #201

**Comment by tatarsky** [comment]
I will likely perform the first step of the backend Lustre server update today: the update of the management server and its IEEL software. 

I believe this can be done without filesystem interruption but I do need to reboot the system and it currently also serves the SGE job spool. I believe the outage will be brief. And I don't show much activity in the jobs. 

If you have something vital going on this morning where the above sounds too risky, please let me know.

**Comment by tatarsky** [comment]
I am about to reboot that system.

**Comment by tatarsky** [comment]
This is now done. The next steps involve one clarification question to Intel on if the filesystem MUST be stopped and unmounted when redundant servers are involved. The release notes are not clear if its a hard requirement in that case so I have asked for elaboration. 

I have confirmed that CentOS 7.2 clients are fully supported in version 2.4.2.7 and am now grabbing that distro ISO for prep for additional tests. I may ask for one of the new nodes to be used for that test so please be sure to check on #201 first in case we run into severe migration needs at the userspace level.

**Comment by hirokomatsui** [comment]
That sounds good. I see cn12 is running fine with Centos 7.1

**Comment by tatarsky** [comment]
Yes, although thats not really a Lustre test. The old nodes are not on the 10G. As soon as I get one more Intel answer I will probably offline Node 14 or something (new nodes) for a CentOS 7.2 test. 

cn12 I will likely move to CentOS 7.2 shortly just to be consistent. 

Its main purpose is a place for people to ssh to per #201 and see what level of impact the upgrade would create for their modules and code.

**Comment by tatarsky** [comment]
My estimate for downtime IF Intel confirms the filesystem must be shutdown for upgrades regardless of redundancy is a morning (4 hours). I can schedule it early my time if that becomes the case but all users and jobs and notebooks would have to exit. Again, this is an estimate not yet confirmed to be needed for possible discussion.

**Comment by hirokomatsui** [comment]
Probably in your morning time will be better for shutting down. Please let me know as you schedule, I'll make sure to let the people know.

**Comment by tatarsky** [comment]
Hoping Intel says "not really needed for redundant configs" like we have.

**Comment by tatarsky** [comment]
Intel has provided me a procedure but with some caveats that they consider the "shut down the filesystem" a more supported and safer course of action than doing it in redundant pairs.

I will read it again in the morning to see more fully if I agree, but I think given we don't really have backups we should probably stick with the safer course of action for this step is the way I'm reading it tonight. 

So folks might mull a good morning in the next few days.

**Comment by tatarsky** [comment]
My "coffee" mulling is it would be wise given support and lack of backups to do this backend upgrade the officially supported way. That requires filesystem un-mount.

I can do this anytime in the next eight days I will say. After that I am going to go on a vacation (July)

I estimate a morning of complete downtime. (everyone off the systems, all jobs exitted, all notebooks closed)

**Comment by tatarsky** [comment]
Also during my IEEL upgrade R&D I noted this will be our last "release" probably from Intel itself. The article mentions a release in May 2017 but I've not see signs of it. 

https://www.theregister.co.uk/2017/04/18/intel_loses_its_lustre_bins_ownbrand_hpc_filesystem/

So we won't be paying support on Lustre either next year. 

So I'd like to get this nice and stable and in the future it will be the community path we take.

**Comment by tatarsky** [comment]
For my reference: I am monitoring this Github going forward as they claim said release will end up there. 
https://github.com/intel-hpdd

**Comment by hirokomatsui** [comment]
What time in the morning are you available to deal with this? Do you have any estimate time that the file system will be unmounted?

**Comment by tatarsky** [comment]
I can get up nice and early here at 6:00AM or something which is 4:00AM your time. 

My estimate remains I would like 4 hours of downtime to complete the tasks I see outlined. (Basically four CentOS 6.6 -> 6.8 updates and the Lustre bump)

Assuming all goes well. I do not guarantee such a downtime but the steps I see seem thats a reasonable time frame. 

So it would return in theory around 10:00AM my time which is 8:00AM your time.

Make sense? I figure the morning is more desirable for folks due to the time zone differences.

**Comment by tatarsky** [comment]
Note clearly: none of this is NODE level changes. Aka this is just the backend Lustre folks. CentOS 7 work requires a gradual process of node and head node changes. This is to give us a good solid Lustre foundation and complete the ability for having CentOS 7.2 clients.

**Comment by tatarsky** [comment]
I am willing to risk the "live" version BTW if this can't be done. But I would encourage the safer path due to the lack of actual backups. Its not worth a four hour window in my opinion to risk that much data. Given the long uptimes we really need a maintenance window more frequently in the future.

**Comment by hirokomatsui** [comment]
Do you mean it's still worth to take four hours specifically for this update?

**Comment by tatarsky** [comment]
I feel four hours is a very short downtime, yes. Well worth the lower risk and official Intel "blessing" that its done the supported way if there was trouble.

**Comment by tatarsky** [comment]
We BTW crossed 600 days of uptime on many systems today. Thats a long time ;)

**Comment by hirokomatsui** [comment]
Will you or Intel recommendation come up with any safer way in our situation of no backup if we have longer time frame?

**Comment by tatarsky** [comment]
Nope.

**Comment by tatarsky** [comment]
The main thing to remember is this is a patch upgrade (we are remaining 2.X Lustre). Not a major version jump. So I am reasonably confident it should go well. But with ALL filesystems of any complexity there is always some chance of unexpected problems. And for that reason backups are always recommended. From basic disks up to the most complicated configs.

If you would like to pause this process until we have better backups, that is also fine with me. See #200

**Comment by tatarsky** [comment]
One item of note actually the secondary MDS server burped last night and was offline. (Its done this I believe twice in 600+ days and is one of the reasons I seek to update). 

So I did the OS portion of the process on it and timed it. It takes 37 minutes. So given there are now three left I feel my 4 hour estimate is sound and contains room for "issues". 

Note that major issues would require me to contact Intel and that could easily extend out of the window. 

But overall I remain pretty confident in the details of this from the prep work.

**Comment by tatarsky** [comment]
The backend is now CentOS 6.8 updated and with the latest 2.X patch release deployed. I feel this makes things more stable for proceeding with node CentOS 7 conversions. Process was a bit exciting to jump multiple versions. I think perhaps every six months we should review the patch level of Lustre going forward rather than > 600 days.

**Comment by tatarsky** [comment]
OK. I think I have done this right but please test first and then I will make it available from the Jupyterhub config for testing per a @billgreenwald email request.

There is a `c7.q` with currently only one non-high speed old node in it: `cn12`

This is to test scheduler based pipeline items on CentOS 7 in at least some way.
```
qsub -l c7 somescript.sh
```

--or---

`qlogin -l c7`

Remember not much I/O to that node. This list of nodes may expand as we move forward.

**Comment by billgreenwald** [comment]
If you are simply asking for a test that we can qsub to the c7 queue, I just tested a quick echo to a file and it worked.  ie i did 
```
echo " echo 'test' > test.test " | qsub -l c7 -cwd
```
and it made the file as expected and went to the correct queue

If you need a more in depth test, could you clarify what you are looking for?

**Comment by tatarsky** [comment]
Nope thats good enough (I did some others) and also wanted to "@" you for a coordination of the Hub add. What works for you on that addition we discussed? I show several hubs running and it will require a hub restart....

**Comment by billgreenwald** [comment]
I will check with people; potentially bringing it down tomorrow morning before people get in should work, but I will post again and confirm.

**Comment by tatarsky** [comment]
Sure. I have the config line added so basically next convenient restart....

**Comment by tatarsky** [comment]
(and thanks for coordinating the effort)

**Comment by billgreenwald** [comment]
Should be good to do in the morning

**Comment by tatarsky** [comment]
I will do so and do some tests myself! Thank you!

**Comment by tatarsky** [comment]
Hub restarted. Tested only the Python 3 notebook in my environment that uses the source built python. Others may require testing and additions to system python on that OS. (system python is version 2.7 in CentOS 7). But I believe most folks are using their own python (anaconda) and thus their own iPython engines. So just advise if you think system changes are needed or other debug required!

We can add some more C7 nodes if it helps.

**Comment by tatarsky** [comment]
I propose converting based on conversations in #201 the `fl-n-1-1` node to CentOS 7. It would be removed from the current "regular" queues and added to the test "c7" queue.

The purpose is to speed pipeline tests on a 10G connected system. We can then discuss after those continued valuable tests the rate of final cutover and the choice between a single SGE instance and a split one off a converted head node. (both have some merits...best discussed after pipeline unit tests)

**Comment by tatarsky** [comment]
Emailed for update on desire to move a fast node to C7. Noting for status only.

**Comment by hirokomatsui** [comment]
As many people are too busy to test the C7, we are going to take another month, till end of August, for the test. Could you re-schedule the plan for the updating at the beginning of September.
I'm expecting to remake some modules, and will ask you when need help.

**Comment by tatarsky** [comment]
Sounds fine to me. I will mark September as the more likely (still fine to still be busy) start date.

I have some tricks in the module arena when you start thinking about that to "overlay" additional search paths on C7 hosts. (So you don't break the C6 usage during that process)

We can work on that anytime with the `cn12` node. Just let me know and have a great Friday!

**Comment by hirokomatsui** [comment]
I talked with Kelly, and we want to start updating processing next week. Are you available from Monday morning? Could you clarify the updating steps for us.
We would like to use the system during the updating, but some nodes can be off. Jupiterhub and SGE (we can modify the queue if needed) should keep running.
In my understanding, you'll update Luster first, then the head nodes and compute nodes?

**Comment by tatarsky** [comment]
Sure, I can send an email with the steps. Remember we already did the Lustre work back a ways. We are sticking with Lustre 2.3 on CentOS 6 backend systems. 3.0 was basically a full reinstall process. 

So its mostly head node and node migrations to C7. Which can be phased in various ways and I'll get the Kickstart settings ready. Mostly I believe we should start with _one_ of the 10G connected nodes to validate the process.

**Comment by tatarsky** [comment]
Email sent to previous group of people involved.

**Comment by tatarsky** [comment]
`fl-n-1-12` disabled in SGE. No jobs were on it. I will update it to C7 shortly.

**Comment by tatarsky** [comment]
Working on this today. (the node kickstart)

**Comment by tatarsky** [comment]
Migration of the C6 kickstart config for the newer nodes complete. Working on any puppet extras. Lustre mounting (main reason for this 10G connected test to confirm)

**Comment by tatarsky** [comment]
Node allowing logins for basic tests and I'll add it to the c7 queue (and remove from others) before I mess with SGE.

@hirokomatsui do you want that same set of -devel stuff on this one for faster rebuilds than `cn12`? Just let me know. Tracking node config closely in puppet to speed replication of the next ones we decide on.

**Comment by hirokomatsui** [comment]
You mean to install the same -devel packages as the current compute nodes have?

**Comment by tatarsky** [comment]
No, I believe at one point you added some `-devel` to `cn12` in order to help you build some items that were CentOS 7 specific. If you wanted the same items, I can add them to the puppet node default list for CentOS 7.

If that was just a one time thing, normally such things would go on the head node. (Which we don't have yet for CentOS 7)

**Comment by tatarsky** [comment]
I have done I believe the steps to add `fl-n-1-12` to the `c7.q` but I'm not going to enable it for a moment as I see several thousand jobs waiting and in case I made an error I don't want them to run on that machine. 

If you wish to test items until I make that change you are welcome to ssh directly to the system.

**Comment by hirokomatsui** [comment]
I won't need devel packages on compute nodes once all the head and compute nodes run on the same OS. We'll make all the module on the head node basically. Thank you!

**Comment by tatarsky** [comment]
Understood. So basically then the next question which we can discuss is "do you wish more systems converted to C7" or do you wish to continue to test some more.

**Comment by hirokomatsui** [comment]
Yes, I don't think we have to test more.
Please let me know if we have to leave from compute or head nodes for the update.

**Comment by tatarsky** [comment]
I think we need a chat to understand that question and its timing. 

Basically I can upgrade systems at any time but there will of course be impact as I change in particular the head nodes.

I think what would make solid sense is to update `fl-hn2` to CentOS 7 so that

1. I can bring up another Jupyter Hub there to make sure it works OK first before we plan on doing 'fl-hn1`

2. People can use that as a compile host and use with the new CentOS 7 systems on the same SGE config. 

But the main question for you folks combined is do you want to have a set of extra queues to deal with the different OS version nodes while we go or just go quickly in the conversion?

Aka something like:

`c7all.q c7long.q c7short.q c7week.q` 

Or just accept that the all/long/short/week queues will have a decreasing number of C6 nodes in them.

Make sense? Happy to call in or something if it doesn't.

**Comment by tatarsky** [comment]
Also to make it clear: It takes about an hour to kickstart a node to C7. And I can do several at a time. 
So they can be converted quickly if thats easier.

**Comment by hirokomatsui** [comment]
I understood. I think it's easier on the phone. Please call me in my office, 858-822-1029, or let me know your good time to call.

**Comment by hirokomatsui** [comment]
Everyone's off from flh2. Please go ahead!

**Comment by tatarsky** [comment]
Ok. I need to do some analysis of that machine first to make sure I preserve any needed items but I will block logins to it. I am going to mark this effort with another Git issue so people know when its safe to come back.

**Comment by tatarsky** [comment]
Enabled `fl-n-1-12` in the `c7.q` if you need to test, recommend:
```
qlogin -l c7
Your job 3218237 ("QLOGIN") has been submitted
waiting for interactive job to be scheduled ...
Your interactive job 3218237 has been successfully scheduled.
Establishing builtin session to host fl-n-1-12.local ...
fl-n-1-12 [~]$
```

**Comment by tatarsky** [comment]
Per #216 I am now running C7 on the second head node. I will be looking today at running an additional Jupyter Notebook hub on that server. That may take me a bit. It is important to do however before we plan the upgrade of `fl-hn1`. I will provide status as I look at it. I am debating between using the exact same module code or a fresh one with the latest.

**Comment by tatarsky** [comment]
Having trouble with Notebook proxy cookie authentication I suspect because of some shared component. I am trying very hard to not interfere with running notebooks on fl-hn1 but I stress again if you are in the middle of something major please advise as this is a bit complex to deal with and I could cause issues.

**Comment by hirokomatsui** [comment]
Is there any possibility that we have to shut flh1 down, or any services such as SGE? I'll announce people if so. Also please let me know if I need to look up some.

**Comment by s041629** [comment]
I have a lot of jobs running on the opt queue. Will they be affected in any way? They should be done by the weekend

**Comment by tatarsky** [comment]
When we get to upgrading it yes. But at the moment I cannot get the Hub working on CentOS 7 which prevents that upgrade.

**Comment by tatarsky** [comment]
@s041629 nothing will change on the current SGE until I get the JupyterHub code working on CentOS7. As I cannot shutdown and re-install fl-hn1 until I get that confirmed to be doable under CentOS7.

**Comment by tatarsky** [comment]
8 mode nodes will be converted to CentOS 7 this afternoon and accessible via the separate SGE instance on `fl-hn2`. Hub is also back up for testing against that environment which is currently two nodes but more are coming. 

The target 8 nodes are disabled in the other SGE environment on fl-hn1 and no items seen on them.

**Comment by hirokomatsui** [comment]
I made a backup of mysql data, will let everyone move to flh2 as it's ready.

**Comment by tatarsky** [comment]
Sounds good. Its certainly testable with the one node and I'm working on the others.

**Comment by tatarsky** [comment]
The requested 8 nodes are ready on `fl-hn2`. I am double checking the queue config and hub configs for that environment but report any issues here and I'll take care of it.

**Comment by tatarsky** [comment]
C6 Nodes without active jobs have been disabled and are being converted to C7.

Please consider exit of any fl-hn1 based Jupyter Hub jobs as you can to speed the process of migration. 
The hub is available on `fl-hn2` for similar work via Git #217 under C7.

**Comment by tatarsky** [comment]
Closing as primary environment is CentOS 7. Feel free to re-open if missing item but check for existing Git tickets first.
